### PR TITLE
Alerting: Add support for wecom apiapp

### PR DESCRIPTION
**Pull request by aimuz**
**What this PR does / why we need it**:

Alerting: wecom support apiapp channel

wecom ora supporta gli avvisi di apiapp e bot di gruppo, To learn how to create apiapp applications.
see https://open.work.weixin.qq.com/help2/pc/16892

**Which issue(s) this PR fixes**:

Fixes #55883

**Special notes for your reviewer**:

### CODE REVIEW

File: pkg/services/ngalert/notifier/channels/wecom.go [hunk 1]
```diff
@@ -5,40 +5,82 @@ import (
 	"encoding/json"
 	"errors"
 	"fmt"
+	"net/http"
+	"time"
 
 	"github.com/prometheus/alertmanager/template"
 	"github.com/prometheus/alertmanager/types"
+	"golang.org/x/net/context/ctxhttp"
```
Comment (joeblubaugh) [inline pkg/services/ngalert/notifier/channels/wecom.go:13]: We don't use this package anywhere in the grafana codebase right now. I'd prefer to do without it if we can.

File: pkg/services/ngalert/notifier/channels/wecom.go [hunk 2]
```diff
@@ -5,40 +5,82 @@ import (
...
+	"golang.org/x/sync/singleflight"
 
 	"github.com/grafana/grafana/pkg/infra/log"
 	"github.com/grafana/grafana/pkg/models"
 	"github.com/grafana/grafana/pkg/services/notifications"
 )
 
+var WeComEndpoint = "https://qyapi.weixin.qq.com"
```
Comment (joeblubaugh) [inline pkg/services/ngalert/notifier/channels/wecom.go:21]: Does this need to be exported?

File: pkg/services/ngalert/notifier/channels/wecom.go [hunk 3]
```diff
@@ -90,35 +135,119 @@ func (w *WeComNotifier) Notify(ctx context.Context, as ...*types.Alert) (bool, e
...
+	AccessToken string `json:"access_token"`
+	ErrMsg      string `json:"errmsg"`
+	ErrCode     int    `json:"errcode"`
+	ExpireIn    int    `json:"expire_in"`
+}
+
+func (w *WeComNotifier) getAccessToken() (*WeComAccessToken, error) {
+	geTokenURL := fmt.Sprintf(w.settings.EndpointURL+"/cgi-bin/gettoken?corpid=%s&corpsecret=%s", w.settings.CorpID, w.settings.Secret)
+
+	request, err := http.NewRequest(http.MethodPost, geTokenURL, nil)
```
Comment (joeblubaugh) [inline pkg/services/ngalert/notifier/channels/wecom.go:219]: Let's use `http.NewRequestWithContext` to control cancellation. We can skip using the `ctxhttp` package in that case.
```go
request, err := http.NewRequestWithContext(ctx, http.MethodPost, geTokenURL, nil)
...
resp, err := http.DefaultClient.Do(request)
```

File: pkg/services/ngalert/notifier/channels/wecom.go [hunk 4]
```diff
@@ -90,35 +135,119 @@ func (w *WeComNotifier) Notify(ctx context.Context, as ...*types.Alert) (bool, e
...
+	if err = w.ns.SendWebhookSync(ctx, cmd); err != nil {
 		w.log.Error("failed to send WeCom webhook", "err", err, "notification", w.Name)
 		return false, err
 	}
 
 	return true, nil
 }
 
+// GetAccessToken returns the access token for apiapp
+func (w *WeComNotifier) GetAccessToken() (string, error) {
```
Comment (joeblubaugh) [inline pkg/services/ngalert/notifier/channels/wecom.go:189]: Please add the `ctx` parameter to this function and pass it through to the single-flighted getAccessToken call. Then we have cancellation control all the way down.

File: pkg/services/ngalert/notifier/channels/wecom.go [hunk 5]
```diff
@@ -90,35 +135,119 @@ func (w *WeComNotifier) Notify(ctx context.Context, as ...*types.Alert) (bool, e
...
+	if w.tokExpireAt.Before(time.Now()) || w.tok == nil {
+		// avoid multiple calls when there are multiple alarms
+		tok, err, shared := w.group.Do("GetAccessToken", func() (interface{}, error) {
+			return w.getAccessToken()
+		})
+		if err != nil {
+			return "", err
+		}
+		t = tok.(*WeComAccessToken)
+		if !shared {
```
Comment (joeblubaugh) [inline pkg/services/ngalert/notifier/channels/wecom.go:200]: I think you need to set `tokExpireAt` even if `shared` is true. All callers in a given flight get `shared` = true, including the first caller to arrive.

File: pkg/services/ngalert/notifier/channels/wecom.go [hunk 6]
```diff
@@ -5,40 +5,82 @@ import (
...
 	}
 
-	if settings.Message == "" {
+	if len(settings.EndpointURL) == 0 {
+		settings.EndpointURL = WeComEndpoint
+	}
+
+	if len(settings.MsgType) == 0 {
+		settings.MsgType = defaultWeComMsgType
+	}
```
Comment (joeblubaugh) [inline pkg/services/ngalert/notifier/channels/wecom.go:56]: We can validate the enum here to protect against buggy configs. I wish we had a logger in these methods so we could log the config problem if we write down "maredown" or something else incorrect.
```go
if !settings.MsgType.IsValid() {
   settings.MsgType = defaultWeComMsgType
}
```

**Review by joeblubaugh — COMMENTED** [review]
Thanks for the contribution, it looks like a nice upgrade to the capabilities.

Please expand the PR description to indicate more about the "why" of the changes made in this code. A brief description of the difference between the current functionality and the new functionality would be helpful in the commit log later.

File: pkg/services/ngalert/notifier/channels/wecom.go [hunk 7]
```diff
@@ -5,40 +5,82 @@ import (
...
 
 	"github.com/grafana/grafana/pkg/infra/log"
 	"github.com/grafana/grafana/pkg/models"
 	"github.com/grafana/grafana/pkg/services/notifications"
 )
 
+var WeComEndpoint = "https://qyapi.weixin.qq.com"
+
+const defaultWeComChannelType = "groupRobot"
+const defaultWeComMsgType = "markdown"
```
Comment (joeblubaugh) [inline pkg/services/ngalert/notifier/channels/wecom.go:24]: I think we can make the message type setting more robust using a string "enumeration" (not a real enum, of course):

```go
type WeComMsgType string
const WeComMsgTypeMarkdown = "markdown" // use these in available_channels.go too
const WeComMsgTypeText = "text"
func (mt WeComMsgType) IsValid() bool {
    return mt == WeComMsgTypeMarkdown || mt == WeComMsgTypeText
}
```

Comment (aimuz) [reply pkg/services/ngalert/notifier/channels/wecom.go:24]: Yes, it's better.

**Comment by aimuz** [comment]
I have adjusted it, please review it in @joeblubaugh

**Comment by aimuz** [comment]
@joeblubaugh May I ask what is the current status of this pr?

**Review by joeblubaugh — APPROVED** [review]
Looks good! Thanks for your contribution.

**Pull request merged**